[BACKEND] Add support for out of tree TTIR/TTGIR passes

CRobeck opened the pull request:
Follow on to: https://github.com/triton-lang/triton/pull/8137. This PR coupled with the Triton pass pipeline override adds the ability to call out of tree TTIR and TTGIR MLIR passes. Integration has been done with both the Python and triton-opt interfaces.

Co-authored w/ @plotfi

plotfi commented on CMakeLists.txt:
```diff
@@ -60,6 +62,7 @@ else()
   set(CMAKE_EXE_LINKER_FLAGS_TRITONRELBUILDWITHASSERTS "/debug:fastlink /INCREMENTAL")
   set(CMAKE_MODULE_LINKER_FLAGS_TRITONRELBUILDWITHASSERTS "/debug:fastlink /INCREMENTAL")
   set(CMAKE_SHARED_LINKER_FLAGS_TRITONRELBUILDWITHASSERTS "/debug:fastlink /INCREMENTAL")
+  set(LLVM_BUILD_SHARED_LIBS "0")
```
I can add a comment here, disabled because even in LLVM dll loading on Windows is not supported based on what I saw in their most recent cmake

plotfi commented on test/lib/Plugins/README.md:
```diff
@@ -0,0 +1,7 @@
+# Triton TTIR and TTGIR Out of Tree Plugin Passes
+
```
Added some basic error handling to the API:

```C++

extern "C" {
    enum TritonPluginResult {
        TP_SUCESS = 0,
        TP_GENERIC_FAILURE = 1
    };
}

#define TRITON_PLUGIN_API \
    extern "C" __attribute__((visibility("default"))) TritonPluginResult

// API:
TRITON_PLUGIN_API
tritonAddPluginPass(mlir::PassManager *pm, const char *passName);

TRITON_PLUGIN_API
tritonRegisterPluginPass(const char *passName);

TRITON_PLUGIN_API
tritonEnumeratePluginPasses(uint32_t *passCount, const char **passNames);
```

ThomasRaoux reviewed (commented):
Overall this looks good, few concerns though:
-What are the effects of compiling with shared library?
-One effect I can think of is that both using shared library for MLIR build and being able to dynamically insert passes are going to break the cache mechanism

plotfi commented:
> this technically doesn't break upstream but it's a bit concerning to expose functionality that can break in very subtle way

@ThomasRaoux Would be reasonable to drop the LLVM_BUILD_SHARED_LIBS option from the top level build for this PR (ie cauterize the potential for subtle changes)?

pawelszczerbuk commented on python/triton/runtime/jit.py:
```diff
@@ -682,6 +682,13 @@ def run(self, *args, grid, warmup, **kwargs):
+        # if we need to recompile since the passes that be added are opauqe to the compiler
+        # until they are called
+        if knobs.runtime.add_stages_inspection_hook is not None:
+            self.device_caches.clear()
```
Just a thought, if we would find a way of inspecting/modifying the pass list at runtime, we could also incorporate it in the hash (I think?)

CRobeck replied:
Added a hash of the plugin compile.py src to the kernel cache key. Not exactly sure if the specialization field the right place for it but seemed reasonable enough. It is much more functional now though, which is nice - the same cached versions of a custom compile.py get recognized and picked up.

CRobeck replied:
OK. I think everything should be in place @ThomasRaoux. Thanks for the input, it's much cleaner now!

plotfi commented:
> > > @ThomasRaoux any more thoughts on this? Cache behavior should be sorted out now.
> > 
> > 
> > some more thoughts about caching, let me write down more details tonight or tomorrow. Sorry for the slow pace
> 
> Thanks. One thing that is not covered with the current implementation is if there are conditionals in the hook code to invoke the plugin pass(s) through a runtime ENV. Haven't decided if that is a failure mode worth trying to capture at the moment.

I think we might also want to add a test that exercises what happens if the .so file is malformed or missing symbols too. This might be fine in a follow up PR though.

ThomasRaoux commented on python/triton/runtime/jit.py:
```diff
@@ -687,6 +687,13 @@ def run(self, *args, grid, warmup, **kwargs):
+        if knobs.runtime.add_stages_inspection_hook is not None:
+            passes = inspect.getsource(knobs.runtime.add_stages_inspection_hook)
+            passes_hash = hashlib.sha256(passes.encode("utf-8")).hexdigest()
+            specialization.append(f'("custom_pipeline", {passes_hash})')
```
this code is called for every dispatch, I guess it should be fairly cheap. Did you measure?

ThomasRaoux replied:
fair enough, ignore this part then.

ThomasRaoux reviewed (commented):
I think this looks good overall. Couple more questions on how to make this fit better. Happy to discuss more but I think this is very close.

ThomasRaoux commented on python/triton/compiler/compiler.py:
```diff
@@ -244,6 +245,8 @@ def compile(src, target=None, options=None, _env_vars=None):
     env_vars = get_cache_invalidating_env_vars() if _env_vars is None else _env_vars
     key = get_cache_key(src, backend, options, env_vars=env_vars)
+    if knobs.runtime.add_stages_inspection_hook is not None:
+        key += inspect.getsource(knobs.runtime.add_stages_inspection_hook)
```
> It makes any change in the entire file the plugin pipeline is in trigger a recompile. Not perfect but I think the next step after that would integrating CMake to hash each file of the plugin that goes into the .so an embedding that in the source which feels like overkill for the moment.

I think this is fine, we do hash all triton files and binary so any change to triton invalidate the cache.

What about moving the hash calculation in the plugin itself? That would leave more control to user to have whatever dependency they want?

CRobeck commented:
@ThomasRaoux I think I've addressed all review concerns. Took another pass and did not see anything glaring. No rush though if you want to take another period to review more.

CRobeck commented on python/triton/compiler/compiler.py:
```diff
@@ -244,6 +245,12 @@ def compile(src, target=None, options=None, _env_vars=None):
+        if callable(get_key_fn):
+            key += get_key_fn()
+        else:
+            raise RuntimeError("A get_key function must be implemented in the add_stages_inspection_hook file")
```
Sure. Updated the return values of the hook and removed the inspect. Take a look and let me know what you think.

ThomasRaoux replied:
thanks, can you update the readme?

CRobeck commented on python/triton/runtime/cache.py:
```diff
@@ -296,6 +297,32 @@ def triton_key():
+        TRITON_PASS_PLUGIN_PATH = os.path.abspath(os.environ['TRITON_PASS_PLUGIN_PATH'])
+        all_paths = [TRITON_PASS_PLUGIN_PATH] + get_so_dependencies(TRITON_PASS_PLUGIN_PATH)
+        all_hashes = [hash_file(p) for p in sorted(all_paths)]
+        contents.append(hashlib.sha256("".join(all_hashes).encode()).hexdigest())
```
You're right. Removed.

ThomasRaoux commented on python/triton/runtime/cache.py:
```diff
@@ -296,6 +296,7 @@ def triton_key():
                 break
             libtriton_hash.update(chunk)
     contents.append(libtriton_hash.hexdigest())
+
```
nit remove empty line

ThomasRaoux reviewed (approved):
LGTM, thanks for the great work!
